### PR TITLE
Added optional headers member variable to RestApiClient class

### LINKED ISSUE #534: Add ability to use headers in the RestApiClient tool
- [x] I have read and agree to the [contributing guidelines](https://github.com/griptape-ai/griptape#contributing).

**Is your feature request related to a problem? Please describe.**
There is a limitation with the RestApiClient class where it currently does not support custom HTTP headers. This becomes a problem when interacting with APIs that require headers for authentication, API versioning, or other purposes. Without this feature, the client cannot be used with a wide range of APIs that have these requirements.

**Describe the solution you'd like**
Enhance the RestApiClient class to support custom HTTP headers. This would involve adding a new attribute to the class for storing headers and updating the HTTP methods (get, post, put, patch, delete) to include these headers in their requests. This enhancement will make the client more versatile and capable of interacting with a broader range of APIs, especially those that require authentication or special instructions passed through headers.

**Describe alternatives you've considered**
None. Seemed pretty straight forward.

**Additional context**
This feature would significantly improve the usability of the RestApiClient for various API interactions, especially in environments where authentication or specific request configurations are necessary. Adding this feature aligns with the goal of making RestApiClient a robust and flexible tool for various API interactions.

### PR DESCRIPTION
**Pull request by mgcollie**
- Added support for custom HTTP headers in RestApiClient class. This allows the client to handle APIs that require authentication or specific headers for requests.
- Updated the `delete` method to handle cases where input parameters might be `None`, enhancing the method's reliability and preventing potential runtime errors.

closes #534

### CODE REVIEW

**Comment by ajenyon** [comment]
I've been testing the update today with a local install of SpaceDeck.  API requests are working using curl, and I can obtain an API token using Griptape without issue (this first step doesn't require a header).

Setup:  
Replace griptape/tools/rest_api_client/tool.py with the updated code

Method (with some hard-coded values to speed up testing):
```
artifact_client = RestApiClient(
    base_url="http://127.0.0.1:9666",
    path="/api/spaces/**********/artifacts",
    description="Create an artifact in a space.",
    off_prompt=False,
    request_body_schema=dumps({
        "type": "object",
        "properties": {
            "space_id": {"type": "string"},
            "x": {"type": "number"},
            "y": {"type": "number"},
            "w": {"type": "number"},
            "h": {"type": "number"},
            "r": {"type": "number"},
            "mime": {"type": "string"},
            "description": {"type": "string"},
        },
        "required": ["space_id", "x", "y", "w", "h", "r", "mime", "description", "header"]
    }),
    response_body_schema=dumps({
        "type": "object",
        "properties": {
            "artifact_id": {"type": "string"}
        },
        "required": ["artifact_id"]
    }),
    headers=dumps({
        "type": "object",
        "properties": {
        "x-spacedeck-auth": {"type": "dict"},
        }
    }),
)  
```
Prompt:       
```
ToolkitTask(
       "Create an artifact in the space id **********, x position 1410, y position 1257, width 320px, height 170px.  The header should read X-Spacedeck-Auth: {{ parent_output }}",
        tools=[artifact_client],
     ),
```
Desired output (curl command that works):

`curl -X POST http://127.0.0.1:9666/api/spaces/**********/artifacts -H "Content-Type: application/json" -H "X-Spacedeck-Auth: **********" -d "{\"space_id\": \"**********\",\"x\": 1410,\"y\": 1257,\"w\": 320,\"h\": 170,\"r\": 0}"`

Output from Griptape Terminal (looking promising, though the header appears to be contained within the body):

```
Subtask 7aa0b905f8a14e72995325ff7e50e384
                             Thought: To create an artifact in the space, I need to make a POST request to the API. The request body should contain the space_id, x, y,  
                             w, h, r, mime, description, and header. The header should contain the API token.

                             Action:
                             {
                               "name": "RestApiClient",
                               "path": "post",
                               "input": {
                                 "values": {
                                   "body": {
                                     "space_id": "**********",
                                     "x": 1410,
                                     "y": 1257,
                                     "w": 320,
                                     "h": 170,
                                     "r": 0,
                                     "mime": "image/png",
                                     "description": "Artifact created by API",
                                     "header": {
                                       "X-Spacedeck-Auth": "**********"
                                     }
                                   }
                                 }
                               }
                             }
```

Expected outcome:  

Artifact ID is returned

Actual outcome:

```
ERROR    Subtask 7aa0b905f8a14e72995325ff7e50e384
                             'str' object has no attribute 'items'
                             Traceback (most recent call last):
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\tasks\action_subtask.py", line 76, in run
                                 response = self._tool.execute(getattr(self._tool, self.action_path), self)
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\tools\base_tool.py", line 106, in execute
                                 output = self.run(activity, subtask, preprocessed_input)
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\tools\base_tool.py", line 115, in run
                                 activity_result = activity(value)
                                                   ^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\utils\decorators.py", line 20, in wrapper
                                 return func(self, *args, **kwargs)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\tools\rest_api_client\tool.py", line 126, in post
                                 response = post(url, json=body, timeout=30, headers=self.headers)
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\api.py", line 115, in post
                                 return request("post", url, data=data, json=json, **kwargs)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\api.py", line 59, in request
                                 return session.request(method=method, url=url, **kwargs)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\sessions.py", line 575, in request
                                 prep = self.prepare_request(req)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\sessions.py", line 486, in prepare_request
                                 p.prepare(
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\models.py", line 369, in prepare
                                 self.prepare_headers(headers)
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\models.py", line 489, in prepare_headers
                                 for header in headers.items():
                                               ^^^^^^^^^^^^^
                             AttributeError: 'str' object has no attribute 'items'
```

Workarounds tried:

-Defined the 'header' in 'Tool.py', as well as the my code as a [str] rather than a [dict].  This DID eventually produce an API call with the header separated from the body (maybe by luck), but the same error "AttributeError: 'str' object has no attribute 'items'":

```
                    INFO     Subtask 0ec4dba846a1458699eadab95d99558f
                             Thought: The error message indicates that there's an issue with the input format. The header should be a separate field in the action, not  
                             inside the body. I'll adjust the action accordingly.

                             Action:
                             {
                             "name": "RestApiClient",
                             "path": "post",
                             "input": {
                                 "values": {
                                     "body": {
                                         "space_id": "**********",
                                         "x": 1410,
                                         "y": 1257,
                                         "w": 320,
                                         "h": 170,
                                         "r": 0,
                                         "mime": "image/png",
                                         "description": "New artifact"
                                     },
                                     "header": {
                                         "X-Spacedeck-Auth": "**********"
                                     }
                                 }
                             }
                             }
```

**Comment by mgcollie** [comment]
@ajenyon seems that in my haste I neglected to make changes required for Schema validation. I just pushed a correction.

**Comment by ajenyon** [comment]
Many thanks!

As far as I can tell, the 'headers' request is correctly formatted as a dict:

```
                               "name": "RestApiClient",
                               "path": "post",
                               "input": {
                                 "values": {
                                   "body": {
                                     "space_id": "<space_id>",
                                     "x": 1410,
                                     "y": 1257,
                                     "w": 320,
                                     "h": 170,
                                     "r": 0,
                                     "mime": "image/png",
                                     "description": "Artifact created by AI"
                                   },
                                   "headers": {
                                     "X-Spacedeck-Auth": "<api_token>"
                                   }
```

The changes you made would seem to enforce the data sent to the 'requests' library is also a dict, but I'm still getting this error, suggesting it's being sent as a string:                               

```
          Traceback (most recent call last):
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\tasks\action_subtask.py", line 76, in run
                                 response = self._tool.execute(getattr(self._tool, self.action_path), self)
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\tools\base_tool.py", line 106, in execute
                                 output = self.run(activity, subtask, preprocessed_input)
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\tools\base_tool.py", line 115, in run
                                 activity_result = activity(value)
                                                   ^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\utils\decorators.py", line 20, in wrapper
                                 return func(self, *args, **kwargs)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\griptape\tools\rest_api_client\tool.py", line 137, in post
                                 response = post(url, json=body, timeout=30, headers=self.headers)
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\api.py", line 115, in post
                                 return request("post", url, data=data, json=json, **kwargs)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\api.py", line 59, in request
                                 return session.request(method=method, url=url, **kwargs)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\sessions.py", line 575, in request
                                 prep = self.prepare_request(req)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\sessions.py", line 486, in prepare_request
                                 p.prepare(
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\models.py", line 369, in prepare
                                 self.prepare_headers(headers)
                               File "C:\Projects\griptape\.venv\Lib\site-packages\requests\models.py", line 489, in prepare_headers
                                 for header in headers.items():
 AttributeError: 'str' object has no attribute 'items'
```

Apologies - beyond my current coding ability to put together a fix

**Comment by mgcollie** [comment]
@ajenyon dumps returns a string, not a dict. can you remove the dumps from your headers and try again

**Comment by mgcollie** [comment]
@vasinov, @collindutter & @andrewfrench I noticed the original implementation only allowed for the path variable to be set upon instantiation, so there'd need to be a separate RestApiClient instance for each individual endpoint that any application was going to use. In addition to the changes for adding support for custom headers, I also made a slight modification to the put, patch, post, get and delete methods to allow for optionally passing the path and headers at the call level when desired/needed. This is optional and should be backwards compatible.

**Comment by ajenyon** [comment]
@mgcollie Still learning - thanks!  I updated my code to read:

```
headers=({
        "type": "object",
        "properties": {
        "x-spacedeck-auth": {"type": "dict"}
        },
    }),)
```

'Headers' now does appear to be a dictionary, but this throws a new error:

```
 Subtask a2f9f858080b4fa9ace6d110654689e2
                             Response: Header part ({'x-spacedeck-auth': {'type': 'string'}}) from ('properties', {'x-spacedeck-auth': {'type': 'string'}}) must be of   
                             type str or bytes, not <class 'dict'>
```

I changed the code to read:

```
headers=({
        "type": "object",
        "properties": dumps({
        "x-spacedeck-auth": {"type": "dict"}
        }),
    }),
```

Which does not error, and produces the following API call:

```
 {
                               "name": "RestApiClient",
                               "path": "post",
                               "input": {
                                 "values": {
                                   "body": {
                                     "space_id": "*********",
                                     "x": 1410,
                                     "y": 1257,
                                     "w": 320,
                                     "h": 170,
                                     "r": 0,
                                     "mime": "text/plain",
                                     "description": "New artifact"
                                   },
                                   "headers": {
                                     "X-Spacedeck-Auth": "*********"
                                   }
                                 }
                               }
                             }
```
This is promising, but the header is in the form:

 "X-Spacedeck-Auth": "*********"
 
 Not
 
  "X-Spacedeck-Auth: *********"
  (Note only one set of quotations)
  
  Is it possible to create the header in this format?
  
  Again, apologies for any rookie coding errors - obviously not asking for any support for SpaceDeck.

**Comment by mgcollie** [comment]
@ajenyon  I'm not very familiar with the SpaceDeck api can you share your code that uses headers for authentication successfully that does not use griptape?  Something that only uses requests perhaps? I'd like to examine the syntax.

**Comment by mgcollie** [comment]
as far as I'm aware request headers should only be a dictionary of string key value pairs. not nested dictionaries.

**Comment by ajenyon** [comment]
curl command that works is (buried) a bit higher in this thread:

`curl -X POST http://127.0.0.1:9666/api/spaces/**********/artifacts -H "Content-Type: application/json" -H "X-Spacedeck-Auth: **********" -d "{\"space_id\": \"**********\",\"x\": 1410,\"y\": 1257,\"w\": 320,\"h\": 170,\"r\": 0}"`

Only clear difference I can see is:

`"X-Spacedeck-Auth": "*********"`
(Both in quotes)

vs

`"X-Spacedeck-Auth: *********"`
(one single set of quotes)

**Comment by mgcollie** [comment]
change headers to this (replace YOUR API KEY with yours, also casing matters):

`headers = {"X-Spacedeck-Auth": "YOUR API KEY"}
`

File: griptape/tools/rest_api_client/tool.py [hunk 1]
```diff
@@ -44,20 +46,27 @@ def full_url(self) -> str:
...
     def put(self, params: dict) -> BaseArtifact:
         from requests import put, exceptions
 
-        values = params["values"]
         base_url = self.base_url
-        path = self.path
-        body = values["body"]
+        values = params.get("values", {})
+        path = values.get("path", self.path)
+        body = values.get("body", {})
```
Comment (collindutter) [inline griptape/tools/rest_api_client/tool.py:63]: Are these changes necessary? We can safely assume these values will be present, so no need for the fallback.

File: griptape/tools/rest_api_client/tool.py [hunk 2]
```diff
@@ -44,20 +46,27 @@ def full_url(self) -> str:
                 {% if _self.response_body_schema %}The response body must follow this JSON schema: {{ _self.response_body_schema }}{% endif %}
                 """
             ),
-            "schema": Schema({Literal("body", description="The request body."): dict}),
+            "schema": Schema(
+                {
+                    Literal("body", description="The request body."): dict,
+                    schema.Optional(Literal("headers", description="The request headers.")): dict,
```
Comment (collindutter) [inline griptape/tools/rest_api_client/tool.py:52]: Does it make sense for the LLM to be setting headers?

File: griptape/tools/rest_api_client/tool.py [hunk 3]
```diff
@@ -29,6 +30,7 @@ class RestApiClient(BaseTool):
     request_query_params_schema: Optional[str] = field(default=None, kw_only=True)
     request_body_schema: Optional[str] = field(default=None, kw_only=True)
     response_body_schema: Optional[str] = field(default=None, kw_only=True)
+    headers: Optional[Dict[str, str]] = field(default=None, kw_only=True)
```
Comment (collindutter) [inline griptape/tools/rest_api_client/tool.py:33]: Let's rename to `request_headers` for consistency.

File: griptape/tools/rest_api_client/tool.py [hunk 4]
```diff
@@ -78,22 +87,25 @@ def put(self, params: dict) -> BaseArtifact:
...
 
-        values = params["values"]
         base_url = self.base_url
-        path = self.path
-        body = values["body"]
-        path_params = values["path_params"]
+        values = params.get("values", {})
+        path = values.get("path", self.path)
+        body = values.get("body", {})
+        path_params = values.get("path_params", [])
```
Comment (collindutter) [inline griptape/tools/rest_api_client/tool.py:102]: Same feedback as above. Are these changes necessary?

File: griptape/tools/rest_api_client/tool.py [hunk 5]
```diff
@@ -108,20 +120,27 @@ def patch(self, params: dict) -> BaseArtifact:
...
     )
     def post(self, params: dict) -> BaseArtifact:
         from requests import post, exceptions
 
-        values = params["values"]
         base_url = self.base_url
-        path = self.path
+        values = params.get("values", {})
+        path = values.get("path", self.path)
+        body = values.get("body", {})
```
Comment (collindutter) [inline griptape/tools/rest_api_client/tool.py:137]: Same as above.

File: griptape/tools/rest_api_client/tool.py [hunk 6]
```diff
@@ -150,19 +170,21 @@ def post(self, params: dict) -> BaseArtifact:
...
         from requests import get, exceptions
 
-        values = params["values"]
         base_url = self.base_url
-        path = self.path
-
         query_params = {}
         path_params = []
+        values = params.get("values", {})
+        path = values.get("path", self.path)
```
Comment (collindutter) [inline griptape/tools/rest_api_client/tool.py:177]: Same as above.

File: griptape/tools/rest_api_client/tool.py [hunk 7]
```diff
@@ -181,23 +203,26 @@ def get(self, params: dict) -> BaseArtifact:
                 {
                     schema.Optional(Literal("query_params", description="The request query parameters.")): dict,
                     schema.Optional(Literal("path_params", description="The request path parameters.")): list,
+                    schema.Optional(Literal("headers", description="The request headers.")): dict,
                 }
             ),
         }
     )
-    def delete(self, params: dict = None) -> BaseArtifact:
+    def delete(self, params: Optional[Dict] = None) -> BaseArtifact:
```
Comment (collindutter) [inline griptape/tools/rest_api_client/tool.py:211]: I don't think the `dict = None` was correct to begin with. Let's remove that and keep it as non-`Optional`.

**Review by collindutter — CHANGES_REQUESTED** [review]

**Comment by mgcollie** [comment]
@collindutter I've resolved all the requested changes

File: griptape/tools/rest_api_client/tool.py [hunk 8]
```diff
@@ -55,9 +57,10 @@ def put(self, params: dict) -> BaseArtifact:
         path = self.path
         body = values["body"]
         url = self._build_url(base_url, path=path)
+        headers = values.get("headers", self.request_headers)
```
Comment (collindutter) [inline griptape/tools/rest_api_client/tool.py:60]: `values["headers"]` will never be set since it is not described in the schema for the LLM to use. Though I don't know if we _want_ the LLM setting headers on the fly. Certainly not if they're going to override the user set headers.

What do you think? Do you have a use case in mind?

File: griptape/tools/rest_api_client/tool.py [hunk 9]
```diff
@@ -185,19 +193,20 @@ def get(self, params: dict) -> BaseArtifact:
...
     )
-    def delete(self, params: dict = None) -> BaseArtifact:
+    def delete(self, params: dict) -> BaseArtifact:
         from requests import delete, exceptions
 
         values = params["values"]
         base_url = self.base_url
-        path = self.path
-
+        path = values.get("path", self.path)
```
Comment (collindutter) [inline griptape/tools/rest_api_client/tool.py:201]: `path` is not in the schema so it will not ever be present in `values`. What is the use case you have in mind here?

File: griptape/tools/rest_api_client/tool.py (hunk 8 above)
Comment (mgcollie) [reply griptape/tools/rest_api_client/tool.py:60]: @collindutter values isn't defined in the schema either. maybe I misunderstand how this is supposed to be used, but from my testing with this using the gitlab api, it seemed that there was a 1:1 relationship between RestApiClient and endpoint for the api. I talked about it in this comment from earlier in the PR:  https://github.com/griptape-ai/griptape/pull/535#issuecomment-1872518512

Comment (mgcollie) [reply griptape/tools/rest_api_client/tool.py:60]: the intention was to be able to give an agent a knowledge base about the available endpoints of the api and their usage, and a single RestApiClient that was instantiated for the base url for the api and then be able to optionally pass custom values for headers & path at runtime based on what it learns from the knowledge base.

File: griptape/tools/rest_api_client/tool.py (hunk 9 above)
Comment (mgcollie) [reply griptape/tools/rest_api_client/tool.py:201]: discussed here: https://github.com/griptape-ai/griptape/pull/535#discussion_r1442218071

File: griptape/tools/rest_api_client/tool.py (hunk 8 above)
Comment (collindutter) [reply griptape/tools/rest_api_client/tool.py:60]: `values` is special field [added by Griptape](https://github.com/griptape-ai/griptape/blob/dev/griptape/mixins/activity_mixin.py#L81). Your use-case makes sense, though I fear the LLM would override user set headers.

Maybe it's just one or the other? Default to user set headers (via `self.headers`), and fall back to LLM set headers (via `values["headers"]`).

@vasinov what do you think?

Comment (mgcollie) [reply griptape/tools/rest_api_client/tool.py:60]: @collindutter thank you for the info about values. Also I see your point and agree about the headers, my main intention and goal was for being able to set the path at runtime.

Comment (collindutter) [reply griptape/tools/rest_api_client/tool.py:60]: @andrewfrench what do you think?

Comment (collindutter) [reply griptape/tools/rest_api_client/tool.py:60]: @mgcollie after discussing it with the team, I think we've landed on not allowing the LLM to set headers. These should always be explicitly set by the developer. So we just want:
```python
response = put(url, json=body, timeout=30, headers=self.request_headers)
```

File: griptape/tools/rest_api_client/tool.py (hunk 9 above)
Comment (collindutter) [reply griptape/tools/rest_api_client/tool.py:201]: Same point here, I think it seems risky to have the LLM take guesses at the `path`. I do think we should have better methods for creating a single `RestApiClient` for multiple paths, but that feels out of scope for this PR.

Comment (andrewfrench) [reply griptape/tools/rest_api_client/tool.py:201]: We've experimentally shown, under specific circumstances, that an LLM can build the entire request including the path and schema (and perhaps headers, too), but it's a very specific case and takes quite a few prerequisites and some massaging to get it to work right. I agree with @collindutter here that it's risky for the default behavior of the tool to include sending authenticated `DELETE` requests at guessed paths and that those should be deterministic configurations.

Comment (mgcollie) [reply griptape/tools/rest_api_client/tool.py:201]: @andrewfrench  & @collindutter thank you, totally understand the concerns. will update the pr this weekend and ping you when the changes are ready.

Comment (mgcollie) [reply griptape/tools/rest_api_client/tool.py]: @andrewfrench @collindutter  ping!  :D
